src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -969,51 +969,60 @@ private async ValueTask<HttpResponseMessage> SendUsingHttp11Async(HttpRequestMes
             }
         }
 
-        private async ValueTask<HttpResponseMessage> DetermineVersionAndSendAsync(HttpRequestMessage request, bool async, bool doRequestAuth, CancellationToken cancellationToken)
+        private ValueTask<HttpResponseMessage> DetermineVersionAndSendAsync(HttpRequestMessage request, bool async, bool doRequestAuth, CancellationToken cancellationToken)
         {
-            HttpResponseMessage? response;
+            // The default configuration of SocketsHttpHandler and HttpRequestMessage has requests use HTTP/1.1.
```

Comment:
Out of curiosity, what state machine are we avoiding here? Seems like if these conditions hold, we should be bailing out of both TrySendUsingHttp[2|3]Async synchronously and quickly.

Reply:
That said, there may be some value in just avoiding the synchronous and non-inlineable calls entirely... in which case can we just move the initial checks from those methods here?

Reply:
> Out of curiosity, what state machine are we avoiding here? Seems like if these conditions hold, we should be bailing out of both TrySendUsingHttp[2|3]Async synchronously and quickly.

It means we can `return SendUsingHttp11Async` rather than `await SendUsingHttp11Async`, so we avoid the state machine for DetermineVersionAndSendAsync.

Reply:
Yeah, that makes sense.

I wonder if we could rearrange the code so that TrySendUsingHttp3Async etc doesn't return null, but instead is responsible for falling back to previous versions when necessary. That way we should be totally synchronous until we come across a version that is actually usable, and only create a state machine at that point. And if the only usable one is HTTP1 then we'd only have the state machine for that one.

Reply:
> And if the only usable one is HTTP1 then we'd only have the state machine for that one.

But we'd still have the state machine for that one, e.g. we'd still have a state machine for SendUsingHttp11Async.  Did I misunderstand?